tests/integration/test_task.py

```diff
@@ -80,3 +82,219 @@ def _run(self, input_files):
+        script=diag_script.as_posix(),
+        settings=diag_settings,
+        output_dir=diag_output_dir.as_posix(),
+    )
+    for task0 in task.flatten():
+        task0.priority = 0
+    tasks.add(task)
+
+    return tasks
+
+
+def test_py_diagnostic_run_sequential_task_fails(monkeypatch, tmp_path):
+    """Run DiagnosticTask sequentially with bad Python script."""
+    diagnostic_text = "import os\n\nprint(cow)"
+    tasks = _get_diagnostic_tasks(tmp_path, diagnostic_text, 'py')
+
+    def _run(self, input_filesi=[]):
+        print(f'running task {self.name}')
+
+    monkeypatch.setattr(BaseTask, '_run', _run)
+
+    with pytest.raises(DiagnosticError) as err_mssg:
+        _run_tasks_sequential(tasks)
+    exp_mssg = "diag_cow.py failed with return code 1"
+    assert exp_mssg in str(err_mssg.value)
+
+
+def test_py_diagnostic_run_parallel_task_fails(monkeypatch, tmp_path):
+    """Run DiagnosticTask parallel with bad Python script."""
+    diagnostic_text = "import os\n\nprint(cow)"
+    tasks = _get_diagnostic_tasks(tmp_path, diagnostic_text, 'py')
+
+    def _run(self, input_filesi=[]):
+        print(f'running task {self.name}')
+
+    monkeypatch.setattr(BaseTask, '_run', _run)
+
+    with pytest.raises(DiagnosticError) as err_mssg:
+        _run_tasks_parallel(tasks, 2)
+    exp_mssg = "diag_cow.py failed with return code 1"
+    assert exp_mssg in str(err_mssg.value)
+
+
+def test_py_diagnostic_run_parallel_task(monkeypatch, tmp_path):
+    """Run DiagnosticTask in parallel with OK Python script."""
+    diagnostic_text = "import os\n\nprint('cow')"
+    tasks = _get_diagnostic_tasks(tmp_path, diagnostic_text, 'py')
+
+    def _run(self, input_filesi=[]):
+        print(f'running task {self.name}')
+
+    monkeypatch.setattr(BaseTask, '_run', _run)
+
+    _run_tasks_parallel(tasks, 2)
+
+
+@pytest.mark.install
+def test_ncl_diagnostic_run_parallel_task_fails(monkeypatch, tmp_path):
+    """Run DiagnosticTask in parallel with OK NCL script."""
+    diagnostic_text = "cows on the [river]"
+    tasks = _get_diagnostic_tasks(tmp_path, diagnostic_text, 'ncl')
+
+    def _run(self, input_filesi=[]):
+        print(f'running task {self.name}')
+
+    monkeypatch.setattr(BaseTask, '_run', _run)
+
+    with pytest.raises(DiagnosticError) as err_mssg:
+        _run_tasks_parallel(tasks, 2)
+    exp_mssg_1 = "An error occurred during execution of NCL script"
+    exp_mssg_2 = "diag_cow.ncl"
+    assert exp_mssg_1 in str(err_mssg.value)
+    assert exp_mssg_2 in str(err_mssg.value)
+
+
+@pytest.mark.install
+def test_ncl_diagnostic_run_parallel_task(monkeypatch, tmp_path):
+    """Run DiagnosticTask in parallel with OK NCL script."""
+    diagnostic_text = _py2ncl({'cow': 22}, 'tas')
+    tasks = _get_diagnostic_tasks(tmp_path, diagnostic_text, 'ncl')
+
+    def _run(self, input_filesi=[]):
+        print(f'running task {self.name}')
+
+    monkeypatch.setattr(BaseTask, '_run', _run)
+
+    _run_tasks_parallel(tasks, 2)
```

Comment:
`run_tasks`, `_run_tasks_sequential`, and `_run_tasks_parallel` are already reasonably well tested with the current tests, so I do not think there is a need to test them more. Maybe you could just use the `run` method of the Task object to test running?

Reply:
well they're wrappers of the `task.run()` object anyway no? I can just remove one of the tests so there's no double-testing, what you think?

Reply:
It would be better test smaller amounts of code per test. I know I have not been setting a very good example here, this is because of lack of time. It would probably be better to use the tests of @jvegasbsc as an example of how to write good tests (but do use pytest rather than the unittest.UnitTest class please) than anything I implemented.

Reply:
Good call! am only using pytest, will break them up in smaller bits